[AMDGPU] Prefer lower total register usage in regions with spilling

jrbyrnes opened the pull request:
In general, we say RP is less if it:
1. has better occupancy
2. has same occupancy, but better tuple register pressure
3. has same occupancy, and same tuple register pressure, but better register pressure.

This is good for the general case, but is not good for the case when there is spilling. In cases with spilling, we should prefer the schedule with less RP rather than giving preference to tuple RP. Otherwise, we may increase RP by hundreds in order to save a few tuples.

A note on the test:
For the OccInitialSchedule stage we have
PressureAfter: VGPRs: 67 AGPRs: 0, SGPRs: 8, LVGPR WT: 0, LSGPR WT: 8
PressureBefore: VGPRs: 72 AGPRs: 0, SGPRs: 4, LVGPR WT: 0, LSGPR WT: 4

We have improved VGPR RP at expense of SGPR Tuple pressure. The scheduler decided to define and use the second sgpr_128 before killing the first. We should not prefer the original order.

jrbyrnes committed:
[AMDGPU] Prefer lower total register usage in regions with spilling

Change-Id: I0e4a631a22f5501b1586d1b8fc999eabd572268c

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -115,7 +119,19 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
   }
 
-  // compare large regs pressure
+  // In regions with spilling, we should give prefernce to the schedule with
```
Typo prefernce

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -115,7 +119,19 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+      return GPRPressure < OtherGPRPressure;
+  }
+
+  // Give second prefernce to less register tuple pressure
```
Typo prefernce. s/less/lower/?

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -130,6 +146,8 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
     }
   }
+
+  // Give final prefernce to less general RP
```
Typo prefernce

arsenm commented on llvm/test/CodeGen/AMDGPU/spill-regpressure-less.mir:
```diff
@@ -0,0 +1,353 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 3
+# RUN: llc -march=amdgcn -mcpu=gfx90a -start-before=machine-scheduler -stop-after=machine-scheduler -verify-machineinstrs -o - %s | FileCheck -check-prefix=GCN %s
```
-run-pass=machine-scheduler should work. also don't need -verify-machineinstrs, used -verify-misched to make sure we get a verify run with LiveIntervals

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1184,8 +1184,7 @@ bool ILPInitialScheduleStage::shouldRevertScheduling(unsigned WavesAfter) {
   if (WavesAfter <= MFI.getMinWavesPerEU() &&
-      !PressureAfter.less(ST, PressureBefore) &&
-      isRegionWithExcessRP()) {
+      !PressureAfter.less(ST, MFI, PressureBefore) && isRegionWithExcessRP()) {
```
We should probably decide which kind of register excess is important here and compare schedules accordingly.

jrbyrnes replied:
See above comment

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -115,7 +119,19 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+
+    if (GPRPressure != OtherGPRPressure)
+      return GPRPressure < OtherGPRPressure;
+  }
```
Hi Valery -- 

Thanks for your comments. 

This patch doesn't really intend to answer the question "which schedule to prefer" in general, but, instead, which schedule to prefer when there is spilling.

For this patch, we would like to determine which register type which is excessive / is concerning for spilling. For this purpose, I think the current determination of SGPRImportant is okay. The only default path to GCNRegPressure::less is via `mayCauseSpilling`. GCNRegPressure::less will (now) only be entered if `WavesAfter <= MFI.getMinWavesPerEU() && isRegionWithExcessRP() `. Thus, we know occupancy is concerning, and that either SGPR or VGPR pressure is excessive (I have passed the latter as a flag). The occupancy checks will pick up on which type of register is excessive -- unless there is disagreement between before / after or if they are both excessive.

vpykhtin replied:
You may have excess in SGPRs but not in VGPRs, but both SGPROcc == VGPROcc == 1. In this case it would compare VGPRs (by SGPRImportant = SGPROcc < VGPROcc) but this isn't what you want. 

BTW SGPRs can be spilled into VGPRs, right? How about translating SGPR excess into a number of VGPRs used for SGPR spilling, add that number to the number of already used VGPRs and comparing only VGPRs?

Something like NumVGPRforSGPRSpills = ceil(SGPRExcess / Wavesize)? This is very conservative of course.

I'm not sure how SGPR to VGPR spilling works, cc @arsenm.

jrbyrnes replied:
Hi Valery,

Please see the new iteration of the work -- this has better `SGPRImportant` handling. It builds upon "Improve ExcessRP Flagging". As mentioned below, if this approach is agreed upon (at least at a high level), I'll create a separate PR for that patch.

> How about translating SGPR excess into a number of VGPRs used for SGPR spilling, add that number to the number of already used VGPRs and comparing only VGPRs?

I  think it is best to do this separately, but it does seem like an interesting feature.

jrbyrnes commented:
This work now builds on top of another patch "Improve ExcessRP flagging". That patch alone has functional changes to the scheduler so I will begin to work on putting out a PR for that work, unless there is reason to not do so.

arsenm commented:
> This work now builds on top of another patch "Improve ExcessRP flagging". That patch alone has functional changes to the scheduler so I will begin to work on putting out a PR for that work, unless there is reason to not do so.

That's what you're supposed to do

jrbyrnes commented:
Directly calculate excess pressure in `GCNPressure.less()` to remove patch dependence and decouple. 

We should still separate the `RegionsWithExcessRP` flag in scheduler as it reduces compile time and improves scheduling in regions with spilling, but decoupling is preferred.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,8 +88,8 @@ void GCNRegPressure::inc(unsigned Reg,
 
-bool GCNRegPressure::less(const GCNSubtarget &ST,
-                          const GCNRegPressure& O,
+bool GCNRegPressure::less(const GCNSubtarget &ST, const MachineFunction &MF,
```
can you leave only MF parameter and obtain ST from it, like

`const GCNSubtarget &ST = MF.getSubtarget<GCNSubtarget>();`

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -102,20 +102,56 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+      // If the pressures are the same, fall through to the subsequent checks
+      if (GPRPressure != OtherGPRPressure)
+        return GPRPressure < OtherGPRPressure;
+    }
```
This seems reasonable. So in the case when both sides has SGPR and VGPR excess we prefer SGPR?
What if both sides have equal SGPR excess, should we compare VGPR excess then, or vice versa?

jrbyrnes replied:
I implemented the weight cost comparison, with a very limited cost model. With the inclusion of additional types of register excess (e.g. archVGPR / AGPR) it doesn't make sense to try to compare just one. The new implementation should be able to choose the better RP when there are excesses from multiple types.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -102,20 +102,56 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+  bool ExcessVGPR = getVGPRNum(false) > MaxVGPRs || getAGPRNum() > MaxVGPRs;
+  bool ExcessSGPR = getSGPRNum() > MaxSGPRs;
+  bool OtherExcessVGPR =
+      O.getVGPRNum(false) > MaxVGPRs || O.getAGPRNum() > MaxVGPRs;
```
Should _getVGPRNum_ has _ST.hasGFX90AInsts()_ here?

arsenm replied:
Is comparing getAGPRNum to MaxVGPRs correct? They'd be different for gfx90a

jrbyrnes replied:
Thanks for pointing this out -- 

These conditions were reformulated from existing code in `checkScheduling`

```
  unsigned MaxVGPRs = ST.getMaxNumVGPRs(MF);
  unsigned MaxSGPRs = ST.getMaxNumSGPRs(MF);

  if (PressureAfter.getVGPRNum(false) > MaxVGPRs ||
      PressureAfter.getAGPRNum() > MaxVGPRs ||
      PressureAfter.getSGPRNum() > MaxSGPRs) {
    DAG.RescheduleRegions[RegionIdx] = true;
    DAG.RegionsWithHighRP[RegionIdx] = true;
    DAG.RegionsWithExcessRP[RegionIdx] = true;
  }
 
```

This appears to be wrong in unified register file case. A scheduling region with pressure - 276 VGPR, 0 AGPR is not flagged as having excess with min-waves-per-eu of 1 even though the ISA states "A wave may have up to 512 total VGPRs, 256 of each type". As it stands, it will allow up to 512 VGPR before flagging. I'll address this in a separate patch.

jrbyrnes commented:
Weighted cost + addressable register limits

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,9 +108,62 @@ void GCNRegPressure::inc(unsigned Reg,
+        ((OtherExcessVGPR + OtherExcessArchVGPR + OtherExcessAGPR) -
+         (ExcessVGPR + ExcessArchVGPR + ExcessAGPR));
+    uint64_t VGPRDiffWeight =
+        std::min(VGPRDiffBase, Max.max() / ExcessVGPRWeight) * ExcessVGPRWeight;
```
You're going to divide by zero since ExcessVGPRWeight has range starting from 0

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -18,6 +18,26 @@
+  ExcessVGPRWeightParser(cl::Option &O) : cl::parser<unsigned>(O) {}
+
+  bool parse(cl::Option &O, StringRef ArgName, StringRef Arg, unsigned &Value) {
+    if (Arg.getAsInteger(0, Value) || Value > 2048 || Value < 0)
```
Value is unsigned, are you really going to test if Value is below zero?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,9 +108,62 @@ void GCNRegPressure::inc(unsigned Reg,
+  // lower general RP.
+  if (ExcessRP || OtherExcessRP) {
+    std::numeric_limits<uint64_t> Max;
+    uint64_t VGPRDiffBase =
```
unsigned?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,9 +108,62 @@ void GCNRegPressure::inc(unsigned Reg,
+         (ExcessVGPR + ExcessArchVGPR + ExcessAGPR));
+    uint64_t VGPRDiffWeight =
+        std::min(VGPRDiffBase, Max.max() / ExcessVGPRWeight) * ExcessVGPRWeight;
+    uint64_t SGPRDiffWeight = OtherExcessSGPR - ExcessSGPR;
```
unsigned?
Should it be ExcessSGPR - OtherExcessSGPR?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,9 +108,62 @@ void GCNRegPressure::inc(unsigned Reg,
+    uint64_t SGPRDiffWeight = OtherExcessSGPR - ExcessSGPR;
+
+    if (VGPRDiffWeight + SGPRDiffWeight != 0)
+      return VGPRDiffWeight + SGPRDiffWeight > 0;
```
you're comparing unsigned value here

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -18,6 +18,26 @@
+} // end anonymous namespace
+
+static cl::opt<unsigned, false, ExcessVGPRWeightParser> ExcessVGPRWeight(
+    "amdgpu-excess-vgpr-weight", cl::init(32), cl::Hidden,
```
Is this default value equals wavesize meaning that VGPR spill is 32 times more expensive than SGPR? What about wavesize 64?

jrbyrnes replied:
Thanks again for the thoughts @vpykhtin  -- I've let it default to `ST.getWavefrontSize()` , the rationale being that spilling a VGPR is equivalent to spilling WavefrontSize SGPRs, but this is a simple heuristic -- thus it is a configurable parameter.

vpykhtin replied:
I don't have experience on spilling but I think "VGPR is equivalent to spilling WavefrontSize SGPRs" isn't even theoretical maximum because we have to feed 32/64 SGPRs to workitems first. Another question is how efficient the VGPR spills are because in the worst case we would spill a VGPR to take a single SGPR from it. I don't have a good suggestion on default value though.

dmikushin commented:
@jrbyrnes , is it possible to make the patch to choose the default value of `ExcessVGPRWeight` such that the original behavior of unpatched code could be preserved? It should initially minimize unexpected effects in the high-level code and the chance of this code being blamed. Then, you could tune the parameter in a separate update.

jrbyrnes commented:
> is it possible to make the patch to choose the default value of `ExcessVGPRWeight` such that the original behavior of unpatched code could be preserved? 

Unfortunately, it is not possible in the current formulation. The new weighted-cost logic is inconsistent with the unpatched tiered-comparisons logic; mainly because the weight-cost formula does not include all of the terms (e.g. tuple pressure).

dmikushin commented:
@jrbyrnes , since SGPR spills to VGPR, the value of 32 makes sense, as well as 64 for wave64. But why the max value of ExcessVGPRWeight is 2048?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -88,9 +108,68 @@ void GCNRegPressure::inc(unsigned Reg,
                           unsigned MaxOccupancy) const {
+  const GCNSubtarget &ST = MF.getSubtarget<GCNSubtarget>();
+
+  unsigned MaxVGPRs = ST.getMaxNumVGPRs(MF);
```
I remember at some point you had a guard for your checks if any side at lowest possible occupancy? We're calculating the occupancy later anyway so may be it worth to check it before diving into excess magic, just a performance consideration.

jrbyrnes commented:
Updated review to reflect some ideas contained in offline discussion.

The primary goal of the weighted cost calculation is to determine which RP results in less excess VGPR since these will need to be spilled to memory. Since SGPR can be spilled into VGPR, we can reformulate the excess SGPR in terms of VGPR pressure as so: 1 VGPR = WavefrontSize Excess SGPR.

There are some subtlties to consider, however. 1. [0:Wavefrontsize] Excess SGPR are all the same in terms VGPRs reguired (1) -- in other words 1 ExcessSGPR uses the same number of VGPR as 31 ExcessSGPR, 2. N excess SGPR is worse than N - 1 excess SGPR even if they require the same number of VGPRs due to spill code inserted (write/readlane), 3. (2 Excess VGPR, 0 Excess SGPR) is preferable to (1 Excess VGPR, 1 Excess SGPR) since the latter needs to insert more spill code.

Subtlties 2 and 3 can also be thought of as one RP being less than the other.


We should have the following behavior (ExcessVGPR, ExcessSGPR) (WavefrontSize is 64):

|Case		|Pressure 1		|Pressure 2			|Winner     |
|-----------|---------------|-------------------|-----------|
|0			|(2,0)			|(1,1)				|Pressure 1 |
|1 			|(1,1)			|(1,63)				|Pressure 1 |
|2 			|(2,0)			|(1,63)				|Pressure 1 |
|3			|(2,0)			|(1,64)				|PRessure 1 |
|4			|(3,0)			|(1,64)				|PRessure 2 |
|5			|(3,0)			|(1,128)			|PRessure 1 |
|6			|(3,0)			|(0,1280)*			|PRessure 2 |


*VGPR usage is not excess even with VGPR spills from SGPRs.

It is a bit hard to write lit tests for all of these conditions, but I have unit tested the logic and validated the above cases.

Also,Removed the flag to set the relative weighting of excess SGPR and VGPR since the current formulation only really makes sense if using the WavefrontSize.

jrbyrnes commented:
> @jrbyrnes , since SGPR spills to VGPR, the value of 32 makes sense, as well as 64 for wave64. But why the max value of ExcessVGPRWeight is 2048?

I set that to the upper limit to avoid overflow concerns, and I couldn't imagine any reason why we would want to set it any higher than that. But, in the latest iteration, I have removed the option altogether.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -104,18 +105,82 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+    // If excess VGPR is the same, prefer the lower excess SGPR
+    if (VGPRDiff != 0)
+      return VGPRDiff > 0;
+    else if (SGPRDiff != 0)
```
Just a nit: 'else' isn't needed due to the return above. SGPRDiff can be calculated after VGPRDiff != 0 is tested.

jrbyrnes commented:
Hi @arsenm , does the latest iteration of this PR resolve your change request? Also, do you have a comment on the excess RP calculation for the unified register file case (see Valery's latest comment). Thanks.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -104,18 +105,86 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
   }
 
-  // compare large regs pressure
+  // Give thid precedence to lower register tuple pressure.
```
Typo 'thid'

jrbyrnes commented:
> This part doesn't entirely make sense to me. The SGPR spill code may end up larger, but could still be faster. Unless the spill-to-vgpr handling ends up inducing a VGPR spill, I'd spill expect to prefer an excess SGPR to excess VGPR

Okay, I have modified it to reflect the principle: prefer the register pressure with SGPR-spills-to-VGPR when there is a tie in Excess VGPR pressure.

That the following behavior is observed:


|Case		|Pressure 1		|Pressure 2			|Winner     |
|-----------|---------------|-------------------|-----------|
|0			|(2,0)			|(1,1)				|Pressure 2 |
|1 			|(1,1)			|(1,63)				|Pressure 1 |
|2 			|(2,0)			|(1,63)				|Pressure 2 |
|3			|(2,0)			|(1,64)				|Pressure 2 |
|4			|(3,0)			|(1,64)				|Pressure 2 |
|5			|(3,0)			|(1,128)			|Pressure 2 |
|6			|(3,0)			|(0,1280)*			|Pressure 2 |


It also has the following behvaior: 

|Case		|Pressure 1		|Pressure 2			|Winner     |
|-----------|---------------|-------------------|-----------|
|7			|(0,0)			|(0,1)*				|Pressure 1 |


*VGPR usage is not excess even with VGPR usage from SGPR spills.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -104,18 +105,104 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
+    if (VGPRDiff != 0)
+      return VGPRDiff > 0;
+    if (SGPRDiff != 0) {
+
```
Extra blank line

jrbyrnes commented:
https://github.com/llvm/llvm-project/commit/113052b2b022c4ce45c8003057ae4297d48ed843
